strings/utf8: unneeded copy in validate_utf8

joe-redpanda opened the pull request:
boost::locale::conv::utf_to_utf returns a re-encoded string which is
promptly discarded. This change performs the inner utf8 validation
without storing an additional string.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26653) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26653) before we can accept your contribution.<br/><hr/>**joe-redpanda** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=26653) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26653) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26653) before we can accept your contribution.<br/><hr/>**joe-redpanda** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=26653) it.</sub>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4249a2fd-97a3-4f39-932c-07f3fa7343fd-->
<!--TEST RESULTS-->
<!--b9cb5acc-ac7f-4f52-b5c4-4657596b0f16-->
<!--TEST RESULTS-->
<!--0623bf3f-0aaf-4f66-bda0-791b0d56be6d-->
<!--TEST RESULTS-->
<!--e05f4877-4de9-4740-94a7-348bf68debe5-->
<!--TEST RESULTS-->
<!--c2c7b00a-505b-4c22-9069-6ba9f7c3d6b9-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68289">build#68289</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest           | test_valid_settings        |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68289#0197c860-7731-401e-8748-47ba73b1e378 | FLAKY         | 16/21    | upstream reliability is '85.36585365853658'. current run reliability is '76.19047619047619'. drift is 9.17538 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68289#0197c860-772c-4680-9d00-89394ce49099 | FLAKY         | 19/21    | upstream reliability is '97.43589743589743'. current run reliability is '90.47619047619048'. drift is 6.95971 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68544">build#68544</a></summary>


| test_class      | test_method             | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68544#0197e6ee-d201-4a10-8b10-32abf430f3d5 | FLAKY         | 19/21    | upstream reliability is '99.31818181818181'. current run reliability is '90.47619047619048'. drift is 8.84199 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68595">build#68595</a></summary>


| test_class            | test_method                               | test_arguments                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:------------------------------------------|:-------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68595#0197eaf6-2dc0-4392-92a2-833d483d665a | FLAKY         | 28/30    | upstream reliability is '99.46380697050938'. current run reliability is '90.47619047619048'. drift is 8.98762 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68595#0197eaf6-2dc3-43b5-bce6-3711079c87f2 | FLAKY         | 29/30    | upstream reliability is '99.46236559139786'. current run reliability is '95.23809523809523'. drift is 4.22427 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"}      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68595#0197eaf6-2dc0-4392-92a2-833d483d665a | FLAKY         | 29/30    | upstream reliability is '99.45652173913044'. current run reliability is '95.23809523809523'. drift is 4.21843 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68868">build#68868</a></summary>


| test_class                        | test_method                          | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------|:-------------------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionRecoveryTest            | test_index_recovery                  |                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68868#0197fba5-3320-47e1-9805-e7a00d5ebf80 | FLAKY         | 20/21    | upstream reliability is '98.79518072289156'. current run reliability is '95.23809523809523'. drift is 3.55709 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest             | test_creating_and_listing_migrations |                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68868#0197fba5-3320-43f6-90c7-f0b44adca3fc | FLAKY         | 20/21    | upstream reliability is '99.51456310679612'. current run reliability is '95.23809523809523'. drift is 4.27647 and the allowed drift is set to 50. The test should PASS |
| WriteCachingFailureInjectionTest  | test_unavoidable_data_loss           |                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68868#0197fba9-2c57-4130-9e1c-11f6ae470aff | FLAKY         | 20/21    | upstream reliability is '99.72067039106145'. current run reliability is '95.23809523809523'. drift is 4.48258 and the allowed drift is set to 50. The test should PASS |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets          | {"versions_to_upgrade": 3} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68868#0197fba9-2c5f-4907-843a-6a92ab66d6f9 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69052">build#69052</a></summary>


| test_class             | test_method         | test_arguments           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------|:--------------------|:-------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionRecoveryTest | test_index_recovery |                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69052#0198107f-6085-4b3b-9344-d98567697a59 | FLAKY         | 20/21    | upstream reliability is '99.20212765957447'. current run reliability is '95.23809523809523'. drift is 3.96403 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest     | test_relaxed_acks   | {"write_caching": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69052#01981082-92d4-4553-82df-58993c363cb3 | FLAKY         | 20/21    | upstream reliability is '99.58677685950413'. current run reliability is '95.23809523809523'. drift is 4.34868 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Commit:
strings/utf8: unneeded copy in validate_utf8

boost::locale::conv::utf_to_utf returns a re-encoded string which is
promptly discarded. This change performs the inner utf8 validation
without storing an additional string.

Additionally changes validate_utf8 to call into is_valid_utf8 to
simplify the thrower logic

joe-redpanda commented:
/ci-repeat 10
release
skip-redpanda-build
dt-repeat=10
tests/rptest/tests/node_pool_migration_test.py

bharathv commented on src/v/strings/utf8.h:
```diff
@@ -116,12 +117,19 @@ inline void validate_no_control(std::string_view s) {
+        const boost::locale::utf::code_point c
+          = boost::locale::utf::utf_traits<char>::decode(begin, end);
+        if (
+          c == boost::locale::utf::illegal
```
nit: [is_valid_codepoint](https://beta.boost.org/doc/libs/1_80_0/libs/locale/doc/html/namespaceboost_1_1locale_1_1utf.html#a189805c4e9f4c033ce5f88daf780e08e)

mmaslankaprv commented on src/v/strings/utf8.h:
```diff
@@ -116,12 +117,19 @@ inline void validate_no_control(std::string_view s) {
-          s.data(), s.data() + s.size(), boost::locale::conv::stop);
-    } catch (const boost::locale::conv::conversion_error& ex) {
-        thrower.conversion_error();
+inline void validate_utf8(std::string_view s, Thrower thrower) {
```
Can we take `Thrower` as const ref ?

joe-redpanda replied:
scratch that, is_valid_utf uses a mutating thrower which stores a bool

joe-redpanda replied:
yeah, lets shove the logic into the bool returning method.

joe-redpanda commented:
/dt
